tsl/src/remote/connection_cache.c

```diff
@@ -156,49 +203,226 @@ connection_cache_create()
 	return cache;
 }
 
-Cache *
-remote_connection_cache_pin()
-{
-	return ts_cache_pin(connection_cache_current);
-}
-
 TSConnection *
-remote_connection_cache_get_connection(Cache *cache, TSConnectionId id)
+remote_connection_cache_get_connection(TSConnectionId id)
 {
 	CacheQuery query = { .data = &id };
-	ConnectionCacheEntry *entry = ts_cache_fetch(cache, &query);
+	ConnectionCacheEntry *entry = ts_cache_fetch(connection_cache, &query);
 
 	return entry->conn;
 }
 
-void
-remote_connection_cache_remove(Cache *cache, TSConnectionId id)
+bool
+remote_connection_cache_remove(TSConnectionId id)
 {
-	ts_cache_remove(cache, &id);
+	return ts_cache_remove(connection_cache, &id);
 }
 
 /*
  * Connection invalidation callback function
  *
  * After a change to a pg_foreign_server catalog entry,
- * mark the cache as invalid.
+ * mark the cache entry as invalid.
+ */
+void
+remote_connection_cache_invalidate_callback(Datum arg, int cacheid, uint32 hashvalue)
+{
+	HASH_SEQ_STATUS scan;
+	ConnectionCacheEntry *entry;
+
+	if (cacheid != FOREIGNSERVEROID)
+		return;
+
+	hash_seq_init(&scan, connection_cache->htab);
+
+	while ((entry = hash_seq_search(&scan)) != NULL)
+	{
+		/* hashvalue == 0 means cache reset, so invalidate entire cache */
+		if (hashvalue == 0 || entry->hashvalue == hashvalue)
+			entry->invalidated = true;
+	}
+}
+
+/*
+ * Check if a connection is local.
+ *
+ * This is an imperfect check, but should work for common cases.
+ *
+ * It currently doesn't capture being connected on a local network interface
+ * address. That would require a platform-independent way to lookup local
+ * interface addresses (on UNIX one could use getifaddrs, for instance).
+ */
+static bool
+is_local_connection(const PGconn *conn)
+{
+	const char *host = PQhost(conn);
+
+	return (strcmp("localhost", host) == 0 || strcmp("127.0.0.1", host) ||
```

Comment:
This likely should include at least IPv6 `::1` host as well, but I am also not sure that it is enough as you said.

We probably should set this flag at the moment when host address being resolved first time instead checking it here, we likely will have more context about it from name resolver.